File: app/src/main/java/de/kontranik/freebudget/activity/TransactionActivity.java
package de.kontranik.freebudget.activity;

import android.app.DatePickerDialog;
import android.content.Intent;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.AutoCompleteTextView;
import android.widget.Button;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.Spinner;
import android.widget.RadioButton;

import java.util.Calendar;
import java.util.Date;
import java.util.List;

import de.kontranik.freebudget.R;

import de.kontranik.freebudget.service.SoftKeyboard;
import de.kontranik.freebudget.database.DatabaseAdapter;
import de.kontranik.freebudget.model.Category;
import de.kontranik.freebudget.model.Transaction;

import static de.kontranik.freebudget.activity.CategoryListActivity.RESULT_CATEGORY;
import static de.kontranik.freebudget.service.Constant.TRANS_ID;
import static de.kontranik.freebudget.service.Constant.TRANS_STAT;
import static de.kontranik.freebudget.service.Constant.TRANS_STAT_MINUS;
import static de.kontranik.freebudget.service.Constant.TRANS_STAT_PLUS;
import static de.kontranik.freebudget.service.Constant.TRANS_TYP;
import static de.kontranik.freebudget.service.Constant.TRANS_TYP_PLANNED;

public class TransactionActivity extends AppCompatActivity {

    static final int PICK_CATEGORY_REQUEST = 123;  // The request code

    private EditText editTextDescription;
    private AutoCompleteTextView acTextViewCategory;
    private EditText editTextAmountPlanned, editTextAmountFact;
    private EditText editTextDate;
    private Button buttonDelete, buttonCopy;
    private ImageButton buttonCopyAmount;
    private RadioButton radioButtonReceipts, radioButtonSpending;

    private DatabaseAdapter dbAdapter;
    private long transactionID = 0;

    DatePickerDialog datePickerDialog;
    public int year, month, day;
    public boolean planned = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setTitle(R.string.title_activity_transaction);
        setContentView(R.layout.activity_transaction);

        editTextDescription = (EditText) findViewById(R.id.editText_description);

        editTextDescription.requestFocus();
        SoftKeyboard.showKeyboard(this);

        acTextViewCategory = (AutoCompleteTextView) findViewById(R.id.acTextView_category);
        editTextAmountPlanned = (EditText) findViewById(R.id.editText_amount_planned);
        editTextAmountFact = (EditText) findViewById(R.id.editText_amount_fact);
        buttonDelete = (Button) findViewById(R.id.button_delete);
        buttonCopy = (Button) findViewById(R.id.button_copy);
        buttonCopyAmount = (ImageButton) findViewById(R.id.btn_copy_amount);

        // initiate the date picker and a button
        editTextDate = (EditText) findViewById(R.id.editText_date);

        dbAdapter = new DatabaseAdapter(this);

        dbAdapter.open();
        List<Category> categoryArrayList = dbAdapter.getAllCategory();
        dbAdapter.close();

        ArrayAdapter<Category> adapter = new ArrayAdapter<>(
                this, android.R.layout.simple_dropdown_item_1line, categoryArrayList);
        acTextViewCategory.setAdapter(adapter);
        acTextViewCategory.setOnItemClickListener(new AdapterView.OnItemClickListener() {

            @Override
            public void onItemClick(AdapterView<?> arg0, View arg1, int arg2,
                                    long arg3) {
                // Category selected = (Category) arg0.getAdapter().getItem(arg2);
                /*
                Toast.makeText(RegularTransactionActivity.this,
                        "Clicked " + arg2 + " name: " + selected.getName(),
                        Toast.LENGTH_SHORT).show();
                */
            }
        });

        radioButtonReceipts = (RadioButton) findViewById(R.id.radioButton_receipts);
        radioButtonSpending = (RadioButton) findViewById(R.id.radioButton_spending);

        // perform click event on edit text
        editTextDate.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                // calender class's instance and get current date , month and year from calender
                if ( day == 0 ) {
                    final Calendar c = Calendar.getInstance();
                    year = c.get(Calendar.YEAR); // current year
                    month = c.get(Calendar.MONTH) + 1; // current month
                    day = c.get(Calendar.DAY_OF_MONTH); // current day
                }
                // date picker dialog
                datePickerDialog = new DatePickerDialog(TransactionActivity.this,
                        new DatePickerDialog.OnDateSetListener() {

                            @Override
                            public void onDateSet(DatePicker view, int year,
                                                  int monthOfYear, int dayOfMonth) {
                                // set day of month , month and year value in the edit text

                                setDateBox(year, monthOfYear+1, dayOfMonth);
                            }
                        }, year, month - 1, day);
                datePickerDialog.show();
            }
        });

        Bundle extras = getIntent().getExtras();
        if (extras != null) {
            if ( extras.containsKey(TRANS_ID) ) transactionID = extras.getLong(TRANS_ID);
            if ( extras.containsKey(TRANS_TYP) ) planned = extras.getString(TRANS_TYP).equals(TRANS_TYP_PLANNED);
        }
        //
        long displayDate;

        if (transactionID > 0) {
            // get entry from db
            dbAdapter.open();
            Transaction transaction = dbAdapter.getTransaction(transactionID);
            editTextDescription.setText(transaction.getDescription());

            acTextViewCategory.setText(transaction.getCategory());

            editTextAmountFact.setText(String.valueOf(Math.abs(transaction.getAmount_fact())));
            editTextAmountPlanned.setText(String.valueOf(Math.abs(transaction.getAmount_planned())));

            if (transaction.getDate() > 0) displayDate = transaction.getDate();
            else displayDate = new Date().getTime();

            if (transaction.getAmount_fact() > 0) {
                radioButtonReceipts.setChecked(true);
            } else if (transaction.getAmount_fact() < 0 ) {
                radioButtonSpending.setChecked(true);
            } else if (transaction.getAmount_planned() > 0) {
                radioButtonReceipts.setChecked(true);
            } else if (transaction.getAmount_planned() < 0 ) {
                radioButtonSpending.setChecked(true);
            }

            dbAdapter.close();
        } else {
            String transStat = TRANS_STAT_MINUS;

            if (extras != null) {
                if ( extras.containsKey(TRANS_STAT) ) transStat = extras.getString(TRANS_STAT);
            }

            if (transStat != null && transStat.equals(TRANS_STAT_PLUS)) {
                radioButtonReceipts.setChecked(true);
            } else {
                radioButtonSpending.setChecked(true);
            }

            displayDate = new Date().getTime();

            // hide delete button
            buttonDelete.setVisibility(View.GONE);

            Log.d("NIK", String.valueOf(planned));

        }

        if ( displayDate > 0) {
            Calendar calendar = Calendar.getInstance();
            calendar.setTime(new Date(displayDate));
            this.year = calendar.get(Calendar.YEAR);
            this.month = calendar.get(Calendar.MONTH) + 1;
            this.day = calendar.get(Calendar.DAY_OF_MONTH);
        }
        setDateBox(year, month, day);

        editTextAmountPlanned.setEnabled(planned);
        editTextAmountFact.setEnabled(!planned);
        buttonCopyAmount.setEnabled(!planned);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        // Check which request we're responding to
        if (requestCode == PICK_CATEGORY_REQUEST) {
            // Make sure the request was successful
            if (resultCode == RESULT_OK) {
                acTextViewCategory.setText( data.getStringExtra(RESULT_CATEGORY) );
            }
        }
    }

    private void setDateBox(int year, int month, int day) {
        this.year = year;
        this.month = month;
        this.day = day;
        editTextDate.setText( String.format("%02d/%02d/%04d", day, month, year) );
    }

    // find Index of Spinner by Value
    private int getIndex(Spinner spinner, String myString)
    {
        int index = 0;

        for (int i=0;i<spinner.getCount();i++){
            if (spinner.getItemAtPosition(i).toString().equalsIgnoreCase(myString)){
                index = i;
                break;
            }
        }
        return index;
    }

    public void saveAndClose(View view) {
        save(view);
        SoftKeyboard.hideKeyboard(this);
        this.finish();
        //goHome();
    }

    public void save(View view){

        String description = editTextDescription.getText().toString();
        String categoryName = acTextViewCategory.getText().toString();
        double amountPlanned, amountFact;
        try {
            amountPlanned = Double.parseDouble(editTextAmountPlanned.getText().toString());
            if(radioButtonReceipts.isChecked()) {
                amountPlanned = Math.abs(amountPlanned);
            } else if(radioButtonSpending.isChecked()) {
                amountPlanned = 0 - Math.abs(amountPlanned);
            }
        } catch (Exception e) {
            amountPlanned = 0;
        }

        try {
            amountFact = Double.parseDouble(editTextAmountFact.getText().toString());
            if(radioButtonReceipts.isChecked()) {
                amountFact = Math.abs(amountFact);
            } else if(radioButtonSpending.isChecked()) {
                amountFact = 0 - Math.abs(amountFact);
            }
        } catch (Exception e) {
            amountFact = 0;
        }

        Calendar cal = Calendar.getInstance();
        cal.set(Calendar.YEAR, year);
        cal.set(Calendar.MONTH, month - 1);
        cal.set(Calendar.DAY_OF_MONTH, day);

        dbAdapter.open();

        Transaction entry = new Transaction(
                transactionID, (long) 0, description, categoryName, cal.getTimeInMillis(), amountPlanned, amountFact);

        if (planned) {
            entry.setAmount_planned( amountPlanned );
        } else {
            entry.setAmount_fact( amountFact );
            if ( transactionID > 0) {
                Transaction dbentry = dbAdapter.getTransaction(transactionID);
                if (dbentry != null) {
                    entry.setRegular_id(dbentry.getRegular_id());
                    entry.setAmount_planned(dbentry.getAmount_planned());
                }
            }
        }

        if (transactionID > 0) {
            dbAdapter.update(entry);
        } else {
            dbAdapter.insert(entry);
        }
        dbAdapter.close();

    }

    public void copy(View view){
        transactionID = 0;
        buttonDelete.setVisibility(View.GONE);
        buttonCopy.setVisibility(View.GONE);
    }

    public void delete(View view){
        dbAdapter.open();
        dbAdapter.deleteTransaction(transactionID);
        dbAdapter.close();
        SoftKeyboard.hideKeyboard(this);
        this.finish();
        //goHome();
    }

    public void selectCat(View view) {
        Intent intent = new Intent(this, CategoryListActivity.class);
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP);
        startActivityForResult(intent, PICK_CATEGORY_REQUEST );
    }

    public void copyAmount(View view) {
        editTextAmountFact.setText( editTextAmountPlanned.getText() );
    }
}

File: app/src/main/java/de/kontranik/freebudget/activity/TransactionActivity.kt
package de.kontranik.freebudget.activity

import android.app.DatePickerDialog
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.AdapterView.OnItemClickListener
import android.widget.ArrayAdapter
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import de.kontranik.freebudget.R
import de.kontranik.freebudget.database.viewmodel.CategoryViewModel
import de.kontranik.freebudget.database.viewmodel.TransactionViewModel
import de.kontranik.freebudget.databinding.ActivityTransactionBinding
import de.kontranik.freebudget.model.Category
import de.kontranik.freebudget.model.Transaction
import de.kontranik.freebudget.service.Constant
import de.kontranik.freebudget.service.SoftKeyboard.hideKeyboard
import java.text.DateFormat
import java.util.*
import kotlin.math.abs

class TransactionActivity : AppCompatActivity() {
    private lateinit var binding: ActivityTransactionBinding
    private lateinit var mTransactionViewModel: TransactionViewModel
    private lateinit var mCategoryViewModel: CategoryViewModel

    private var transactionID: Long? = null
    private var regularCreateTime: Long? = null

    private var year = 0
    var month = 0
    var day = 0
    private var planned = false
    private var dbentryAmountPlanned: Double = 0.0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setTitle(R.string.title_activity_transaction)
        binding = ActivityTransactionBinding.inflate(layoutInflater)
        setContentView(binding.root)

        mTransactionViewModel = ViewModelProvider(this)[TransactionViewModel::class.java]
        mCategoryViewModel = ViewModelProvider(this)[CategoryViewModel::class.java]

        binding.editTextDescription.requestFocus()
        //showKeyboard(this)

        binding.buttonCopy.setOnClickListener { copy() }
        binding.buttonSave.setOnClickListener { save() }
        binding.buttonSaveAndClose.setOnClickListener { saveAndClose() }
        binding.buttonClose.setOnClickListener { close() }
        binding.buttonDelete.setOnClickListener { delete() }
        binding.btnCopyAmount.setOnClickListener { copyAmount() }
        binding.btnSelectCat.setOnClickListener { selectCat() }

        val categoryArrayList: MutableList<Category> = mutableListOf()

        val adapter = ArrayAdapter(
            this, android.R.layout.simple_dropdown_item_1line, categoryArrayList
        )
        binding.acTextViewCategory.setAdapter(adapter)
        binding.acTextViewCategory.onItemClickListener =
            OnItemClickListener { _, _, _, _ -> // Category selected = (Category) arg0.getAdapter().getItem(arg2);
                /*
                    Toast.makeText(RegularTransactionActivity.this,
                            "Clicked " + arg2 + " name: " + selected.getName(),
                            Toast.LENGTH_SHORT).show();
                    */
            }

        // perform click event on edit text
        binding.buttonDate.setOnClickListener { // calender class's instance and get current date , month and year from calender
            if (day == 0) {
                val c = Calendar.getInstance()
                year = c[Calendar.YEAR] // current year
                month = c[Calendar.MONTH] + 1 // current month
                day = c[Calendar.DAY_OF_MONTH] // current day
            }
            // date picker dialog
            val datePickerDialog = DatePickerDialog(this@TransactionActivity,
                { _, year, monthOfYear, dayOfMonth -> // set day of month , month and year value in the edit text
                    setDateBox(year, monthOfYear + 1, dayOfMonth)
                }, year, month - 1, day
            )
            datePickerDialog.show()
        }
        val extras = intent.extras
        if (extras != null) {
            if (extras.containsKey(Constant.TRANS_ID)) transactionID = extras.getLong(Constant.TRANS_ID)
            if (extras.containsKey(Constant.TRANS_TYP)) planned =
                extras.getString(Constant.TRANS_TYP) == Constant.TRANS_TYP_PLANNED
        }

        mCategoryViewModel.mAllCategorys.observe(this) {
            categoryArrayList.clear()
            categoryArrayList.addAll(it)
            adapter.notifyDataSetChanged()
        }

        mTransactionViewModel.transactionById.observe(this) {
            regularCreateTime = it?.regularCreateTime
            refreshView(it)
        }

        if (transactionID != null) {
            // get entry from db
            mTransactionViewModel.loadById(transactionID!!)
        } else {
            var transStat: String? = Constant.TRANS_STAT_MINUS
            if (extras != null) {
                if (extras.containsKey(Constant.TRANS_STAT)) transStat = extras.getString(Constant.TRANS_STAT)
            }
            if (transStat != null && transStat == Constant.TRANS_STAT_PLUS) {
                binding.radioButtonReceipts.isChecked = true
            } else {
                binding.radioButtonSpending.isChecked = true
            }
            val displayDate = Date().time
            val calendar = Calendar.getInstance()
            calendar.time = Date(displayDate)
            year = calendar[Calendar.YEAR]
            month = calendar[Calendar.MONTH] + 1
            day = calendar[Calendar.DAY_OF_MONTH]
            setDateBox(year, month, day)
            // hide delete button
            binding.buttonDelete.visibility = View.GONE
        }

        binding.editTextAmountPlanned.isEnabled = planned
        binding.editTextAmountFact.isEnabled = !planned
        binding.btnCopyAmount.isEnabled = !planned


    }

    private fun refreshView(transaction: Transaction?) {
        if (transaction == null) return

        dbentryAmountPlanned = transaction.amountPlanned

        binding.editTextDescription.setText(transaction.description)
        binding.acTextViewCategory.setText(transaction.category)
        if (transaction.amountFact != 0.0) {
            binding.editTextAmountFact.setText(abs(transaction.amountFact).toString())
        }
        if (transaction.amountPlanned != 0.0) {
            binding.editTextAmountPlanned.setText(abs(transaction.amountPlanned).toString())
        }

        binding.radioButtonReceipts.isChecked = ( transaction.amountFact > 0 || (transaction.amountFact == 0.0 && transaction.amountPlanned >= 0))
        binding.radioButtonSpending.isChecked = ( transaction.amountFact < 0 || (transaction.amountFact == 0.0 && transaction.amountPlanned < 0))


        val displayDate = if (transaction.date > 0) transaction.date else Date().time
        if (displayDate > 0) {
            val calendar = Calendar.getInstance()
            calendar.time = Date(displayDate)
            year = calendar[Calendar.YEAR]
            month = calendar[Calendar.MONTH] + 1
            day = calendar[Calendar.DAY_OF_MONTH]
        }
        setDateBox(year, month, day)
    }

    @Deprecated("Deprecated in Java")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        // Check which request we're responding to
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == PICK_CATEGORY_REQUEST) {
            // Make sure the request was successful
            if (resultCode == RESULT_OK) {
                binding.acTextViewCategory.setText(data!!.getStringExtra(CategoryListActivity.RESULT_CATEGORY))
            }
        }
    }

    private fun setDateBox(year: Int, month: Int, day: Int) {
        this.year = year
        this.month = month
        this.day = day
        val cal = Calendar.getInstance()
        cal[year, month - 1] = day
        val df = DateFormat.getDateInstance(DateFormat.SHORT, Locale.getDefault())
        binding.buttonDate.text = df.format(cal.timeInMillis)
    }

    private fun saveAndClose() {
        save()
        close()
    }

    private fun close() {
        hideKeyboard(this)
        finish()
        //goHome();
    }

    private fun save() {
        val description = binding.editTextDescription.text.toString()
        val note = binding.editTextNote.text.toString()
        val categoryName = binding.acTextViewCategory.text.toString()
        var amountPlanned: Double
        try {
            amountPlanned = binding.editTextAmountPlanned.text.toString().toDouble()
            if (binding.radioButtonReceipts.isChecked) {
                amountPlanned = abs(amountPlanned)
            } else if (binding.radioButtonSpending.isChecked) {
                amountPlanned = 0 - abs(amountPlanned)
            }
        } catch (e: Exception) {
            amountPlanned = 0.0
        }
        var amountFact: Double
        try {
            amountFact = binding.editTextAmountFact.text.toString().toDouble()
            if (binding.radioButtonReceipts.isChecked) {
                amountFact = abs(amountFact)
            } else if (binding.radioButtonSpending.isChecked) {
                amountFact = 0 - abs(amountFact)
            }
        } catch (e: Exception) {
            amountFact = 0.0
        }
        val cal = Calendar.getInstance()
        cal[Calendar.YEAR] = year
        cal[Calendar.MONTH] = month - 1
        cal[Calendar.DAY_OF_MONTH] = day

        val entry = Transaction(
            id = transactionID,
            regularCreateTime = regularCreateTime,
            description = description,
            category = categoryName,
            date = cal.timeInMillis,
            amountPlanned = if (planned) amountPlanned else dbentryAmountPlanned ,
            amountFact = amountFact,
            note = note
        )

        if (transactionID != null) {
            mTransactionViewModel.update(entry)
        } else {
            mTransactionViewModel.insert(entry)
        }
    }

    private fun copy() {
        transactionID = null
        binding.buttonDelete.visibility = View.GONE
        binding.buttonCopy.visibility = View.GONE
    }

    private fun delete() {
        transactionID?.let { mTransactionViewModel.delete(it) }
        hideKeyboard(this)
        finish()
        //goHome();
    }

    private fun selectCat() {
        val intent = Intent(this, CategoryListActivity::class.java)
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP)
        startActivityForResult(intent, PICK_CATEGORY_REQUEST)
    }

    private fun copyAmount() {
        binding.editTextAmountFact.text = binding.editTextAmountPlanned.text
    }

    companion object {
        const val PICK_CATEGORY_REQUEST = 123 // The request code
    }
}